### ISSUE TITLE
Suggestion regarding outputs

### ISSUE THREAD

**Issue by sam-lex** [d3lm/ngx-drag-to-select#35]
Hello, thanks for this awesome lib, works perfectly!

I've encountered 1 "challenge" when using this:

I can't get the items from a "add/remove-selection", the only data I can get is the current `selectedItems`. We could do comparisons and get what was added/removed.  
But the problem is that it's not possible to get all items in the "remove-selection" (items that were not already in the selectedItems)

So here are 2 suggestions:  
1. rename current Output `select` to `change`
> (change): Event that is fired whenever the selection changes. The payload ($event) will be the current list of selected items.

2. new Output `select` or `selection` etc..:
> (select): Event that is fired whenever a selection is made. The payload ($event) will contain:
```
{
    items: any[]; // List of items in this selection
    type: 'normal' | 'add' | 'remove'; // Type of selection
}
```

Thank you very much!

**Comment by d3lm** [comment]
Hey @sam-lex, glad you like the lib!

So do I understand correct that when I select items you not only want to get the entire list but also the different, meaning the items that were added / removed once the selection is done? For example:

Selectable items: [1,2,3,4,5,6]
selectedItems: [1]

Now you press a shortcut to extend your selection and select 2,3,4
selectedItems: [1,2,3,4]

In your scenario, you'd also want to get 2,3,4 which would be the difference between the old `selectedItems` and the new ones.

Is that more or less what you want? If not, maybe you can elaborate a little bit more. May I ask what the specific use case is? Just out of curiosity. But in general I think this sounds pretty straightforward and reasonable. Pulling in @timdeschryver on this one too.

**Comment by sam-lex** [comment]
Hi @d3lm, thanks for the reply!

Yes, that's more or less what I want, but not just the difference.
For example:
```
Selectable Items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
selected items: [];
> select (normal): [3,4,5]
--------
selected items: [3, 4, 5]
(emit) [3, 4, 5] // this one is ok, as normal selection will by default clear previous selection
--------
> select (add): [4, 5, 6, 7]
selected items: [3, 4, 5, 6, 7]
(emit) [4, 5, 6, 7] // performing difference with old data would get you only [6, 7]
--------
> select (remove): [6, 7, 8, 9, 10]
selected items: [3, 4, 5]
(emit) [6, 7, 8, 9, 10] // performing difference with old data would get you only [6, 7]
// [8, 9, 10] were not in the selected items before, so is not possible to know about them
```

The reason is because in my case, the selection is persisted somewhere else, normal selection should be additive by default.
One way to solve this would be to create another list that reflects the persisted selection, but this would only solve normal and additive selections, I'd be missing the remove selections.

btw, looking at the source, I noticed that there are two events that does almost the same:
```typescript
  private initSelectedItemsChange() {
    this._selectedItems$
      .pipe(
        auditTime(AUDIT_TIME),
        takeUntil(this.destroy$)
      )
      .subscribe({
        next: selectedItems => {
          this.selectedItemsChange.emit(selectedItems); // <- this
          this.select.emit(selectedItems); // <- this
        },
        complete: () => {
          this.selectedItemsChange.emit([]); // <- only difference, select doesn't emit on complete
        }
      });
  }
```

I have however found a workaround for this 😮, I'll post it in another comment

**Comment by sam-lex** [comment]
Here's my workaround:  
The detailed version is way too long, so I'll just write the logic here:
-> Make default selection: add
-> Make additive selection: remove

1. create dummy selectable item, I call it `#dtsVoid`
```html
<dts-select-container [selectOnDrag]="false" (select)="_onSelect($event)">
  <div #dtsVoid="dtsSelectItem" [dtsSelectItem]="{ isVoid: true }" class="d-none"></div>
```
2. keep selected items initialized like so: `[ dtsVoid ]`
```typescript
// This method should be called once at the beginning
private _initSelection() {
  this.dtsContainer.clearSelection();

  // This is the only way I could get it to select an item programmatically
  // Javascript Wonder: \o/ .selectItem is private
  // If for some reason this method doesn't exist anymore,
  // we will just not be able to use select to remove
  if (!!this.dtsContainer['selectItem'])
    this.dtsContainer['selectItem']( this.dtsVoid );
}
```
3. handle selection
```typescript
_onSelect(selection: any[]) {
  // check if dummy item (#dtsVoid) is in the selection
  if exist
    if selection.length === 1
  //   -> It will get here, when we re-initialize the selection [ dtsVoid ]
  //   -> Or when user used shortcut for removing: so in my case, nothing happens
    else
  //   -> Additive selection: in my case, I use this to interpret as "remove"

  else
  // -> Normal selection: I use this to "add"

  // After we finished processing, re-initialize selection, so it becomes like this again: [ dtsVoid ]
  this._initSelection();
}
```

This is working ok but quite ugly, tonight I'll try checking whether Shift key was pressed during selection, so this way I can minimize my workaround and avoid calling private stuffs 😅

**Comment by timdeschryver** [comment]
I think your use case is a bit too specific to put into this lib?
What about just adding a `onSelect` (name to be determined) when an item gets activated (clicked, dragged, select all, ...), with this you can add the logic you need for your case and if someone else wants to do it a bit different e.g. in your example `performing difference with old data would get you only [6, 7]`, he/she can implement it this way.

With this said I think it would be beneficial to set the selected items from outside the container for example:
```ts
onSelect(item: any) {
  // put some logic here to check if the item can be selected
  this.selected = [...this.selected, item];
}
```
The example above add an item to the selected collection but doesn't show it in the container (ui).

**Comment by sam-lex** [comment]
yes, `@Output() onSelect = new EventEmitter<any>()` would be great!  
Having the new output would make my code cleaner, but that being said, I just got a solution that I'm satisfied with.

So this issue can be closed, unless you'd like to discuss more about it, because in this new solution I make use of `container.selectBoxStyles$`, even though it is public, it is not in the documentation, so I feel like I'm using something that shouldn't be used 😂 

I'll leave my solution below, in case someone finds it useful..

Thank you very much for you help 👍 

1. Bind `[(selectedItems)]` instead, we'll initialize the events manually
```html
<dts-select-container [selectWithShortcut]="true" [(selectedItems)]="_sToggles">
</dts-select-container>
```

2. Initialize events: detect when it started selecting and when it finished selecting
```typescript
private _initSelectionEvents() {
  // Detect when it started selecting
  this.dtsContainer.selectBoxStyles$.pipe(
    takeUntil( this.ngUnsubscribe ),
    // Filtering by opacity is very specific to my use case, my toggles can toggled by themselves...
    // basically this means, if opacity of selectBox is 0, it's just a mouseDown, it's not dragging yet
    // Don't use this filter if you want to capture clicks as well
    distinctUntilChanged((x, y) => x.opacity === y.opacity),
    filter( _ => !!_.opacity ),
    share()
  )
  .subscribe( _ => this._isSelecting = true );

  // Detect when it stopped selecting
  fromEvent(window, 'mouseup').pipe(
    takeUntil( this.ngUnsubscribe ),
    filter( _ => this._isSelecting ),
    share()
  )
  .subscribe((event: MouseEvent) => {
    this._isSelecting = false;
    this._handleSelection( event );
  });
}
```
3. Handle selection
```typescript
private _handleSelection(event: MouseEvent): void {
  // Check whether shift key is pressed
  const isRemove = event.shiftKey;
  // more special cases can be added this way
  const isSpecial = event.altKey && event.ctrlKey && event.metaKey

  // process this._sToggles

  // clear selection
  this.dtsContainer.clearSelection();
}
```

**Comment by timdeschryver** [comment]
@sam-lex while that may work, I would suggest not to do this.
Because the code is coupled to the styling, meaning if we would change the opacity your code will break.

**Comment by d3lm** [comment]
I would also agree that this use case is a little too specific to be implemented in this lib. An output could however make sense for each `dtsSelectItem`, emiting a value when it was selected or deselected.

@sam-lex I would advice against using internal APIs as they might change in the future and that would totally break your code. Just renaming the stream would break your code while not breaking the overall functionality of the lib or introducing breaking changes. `selectBoxStyles$` is only public because it's used inside the template of the container component. This is required to satisfy the AoT compiler. Again, I would not rely on internal APIs and only use the ones that are documented. Everything else might be subject to change.

**Comment by sam-lex** [comment]
oh I see... thanks for the suggestions

I've just successfully replaced `selectBoxStyles$` with `mouseDown + mouseMove` to detect selection like so:

```html
<tbody #dtsContainerHost> <!-- use this element for detecting mouseDown/mouseMove -->
  <dts-select-container>
```
```typescript
@ViewChild('dtsContainerHost') dtsContainerHost: ElementRef;

private _initSelectionEvents() {
  // Detect when user started selecting
  fromEvent(this.dtsContainerHost.nativeElement, 'mousedown').pipe(
    takeUntil( this.ngUnsubscribe ),
    share(),
    switchMap(downEvent => {
      this._mouseDown = true;
      return fromEvent( this.dtsContainerHost.nativeElement, 'mousemove').pipe(
        first(),
        share(),
        map( moveEvent => this._mouseDown )
      );
    })
  )
  .subscribe(isSelecting => {
    this.logger.debug('isSelecting:', isSelecting);
    this._isSelecting = isSelecting;
  });

  // Detect when user is done selecting
  fromEvent(window, 'mouseup').pipe(
    takeUntil( this.ngUnsubscribe ),
    share()
  )
  .subscribe((event: MouseEvent) => {
    this._mouseDown = false;
    if (this._isSelecting) {
      this._isSelecting = false;
      this._handleSelection( event );
    }
  });
}
```

I couldn't attach `#dtsContainerHost` on the `dts-select-container` directly, so I'm attaching it to the parent above it, and works fine.

I've also noticed that there's `SelectContainerComponent.host`, which could be used as well, but I'm assuming that that's also internal, so won't be using that unless told otherwise.

Personally I'm not sure either what outputs would be good for general use, for example:  
- (start): emits when user started drag-select
- (end): emits when drag-select ends

One other thing that I have to do when selecting, is putting a transparent overlay on top of my "selectable items", to avoid sending mouse events to them triggering tooltips etc..

```html
<dts-select-container [selectWithShortcut]="true" [(selectedItems)]="_sToggles">

  <!-- selectable toggles -->

  <div class="transparent-overlay" [hidden]="!_isSelecting"></div>
</dts-select-container>
```

_(having big tooltips popping up when drag-selecting is not good)_
So even if with the new Output, I'd still need to detect when the selection has started to put the overlay.

**Comment by timdeschryver** [comment]
So it boils down to knowing when something was selected and deselected and knowing when the dragging starts?

For the select output I would suggest:

Action                 Output
Click item 1       ->  { item: item1, selected } 
Click item 1       ->  { item: item1, selected } 
Click item 2       ->  { item: item2, selected } 
Ctrl Click item 2  ->  { item: item2, deselected } 
Ctrl Click item 2  ->  { item: item2, deselected } 
Ctrl Click item 3  ->  { item: item3, deselected } 

And the same for the drag actions.
Notice that when even an item was selected it still emits when it gets clicked again.


For knowing when the dragging starts (maybe creating a new issue would be better):
We could also create an Output or do something with `dragging$`?

**Comment by d3lm** [comment]
Yes, I would agree. Having an output on the `dtsSelectItem` that emits whenever the item got selected or deselected is a good idea. Also find it useful for users to know when dragging starts and ends. So we could introduce another output on the `dts-select-container`. That should give you enough flexibility to implement whatever you like to find which items where added or removed in a single selection. This makes it very generic and allows the users of this lib to take it from here and add things on top, things that are more app specific.

**Comment by d3lm** [comment]
Addressed in #51. 

For more outputs on the `SelectContainerComponent` please open a new issue if needed. This probably needs some discussion on when e.g. `dragStart` and `dragEnd` should emit. I don't want to unnecessarily blow up the API of this library. So if you have a valid use case and maybe already a spec in mind how this could work, create an issue and we can continue from there.

**Comment by d3lm** [comment]
@sam-lex Mind to take a look at the PR if this solves your problem? If not, please leave a comment on what is missing.

**Comment by sam-lex** [comment]
@d3lm Hi, PR looks great! 👍 

As for `dragStart`/`dragEnd`, not really necessary I think. I'm using it only to display an overlay when selecting to avoid triggering class changes on my selectable items (tooltips.. popups, etc. which made it buggy - tooltips not closing properly).
I solved it with `mouseDown` + `mouseMove`, this might be specific to my case only.

**Comment by d3lm** [comment]
Ok sweet!

**Comment by d3lm** [comment]
Once the PRs land I ll cut a new release so you can upgrade to the latest version to use all the new features :tada:
